eternalcore-core/src/main/java/com/eternalcode/core/feature/chat/ChatManagerController.java

```diff
@@ -16,52 +19,68 @@
-    permission = { "eternalcore.chat.noslowmode", "eternalcore.chat.bypass" }
+    permission = {"eternalcore.chat.noslowmode", "eternalcore.chat.bypass"}
 )
 @Controller
 class ChatManagerController implements Listener {
 
-    private final ChatManager chatManager;
+    private static final String CHAT_SLOWMODE_BYPASS_PERMISSION = "eternalcore.chat.noslowmode";
+    private static final String CHAT_BYPASS_PERMISSION = "eternalcore.chat.bypass";
+
+    private final ChatService chatService;
+    private final ChatSettings chatSettings;
     private final NoticeService noticeService;
+    private final EventCaller eventCaller;
 
     @Inject
-    ChatManagerController(ChatManager chatManager, NoticeService noticeService) {
-        this.chatManager = chatManager;
+    ChatManagerController(
+        ChatService chatService,
+        ChatSettings chatSettings,
+        NoticeService noticeService,
+        EventCaller eventCaller
+    ) {
+        this.chatService = chatService;
+        this.chatSettings = chatSettings;
         this.noticeService = noticeService;
+        this.eventCaller = eventCaller;
     }
 
     @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
     void onChatSlowMode(AsyncPlayerChatEvent event) {
         Player player = event.getPlayer();
 
-        if (!this.chatManager.getChatSettings().isChatEnabled() && !player.hasPermission("eternalcore.chat.bypass")) {
+        UUID uniqueId = player.getUniqueId();
+
+        if (!this.chatSettings.isChatEnabled() && !player.hasPermission(CHAT_BYPASS_PERMISSION)) {
             this.noticeService.create()
                 .notice(translation -> translation.chat().disabledChatInfo())
-                .player(player.getUniqueId())
+                .player(uniqueId)
                 .send();
 
+            this.eventCaller.callEvent(new ChatRestrictEvent(uniqueId, ChatRestrictCause.CHAT_DISABLED));
+
             event.setCancelled(true);
             return;
         }
 
-        UUID uuid = player.getUniqueId();
-
-        if (this.chatManager.hasSlowedChat(uuid) && !player.hasPermission("eternalcore.chat.noslowmode")) {
-            Duration time = this.chatManager.getSlowDown(uuid);
+        if (this.chatService.hasSlowedChat(uniqueId) && !player.hasPermission(CHAT_SLOWMODE_BYPASS_PERMISSION)) {
+            Duration time = this.chatService.getSlowDown(uniqueId);
 
             this.noticeService
                 .create()
-                .player(player.getUniqueId())
+                .player(uniqueId)
                 .notice(translation -> translation.chat().slowMode())
                 .placeholder("{TIME}", DurationUtil.format(time))
                 .send();
 
+            this.eventCaller.callEvent(new ChatRestrictEvent(uniqueId, ChatRestrictCause.SLOWMODE));
+
             event.setCancelled(true);
         }
     }
 
     @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
     void markUseChat(AsyncPlayerChatEvent event) {
```

Comment:
this is used when, chat is enabled. there is no slowedChat for user. And it fires markUseChat from ChatServiceImpl. Followed in next comment....